### PR TITLE
OSDOCS-1790: Adding auto generated CLI output

### PR DESCRIPTION
**Pull request by bergerhoffer**
https://issues.redhat.com/browse/OSDOCS-1790

Previews:
* Dev commands: https://deploy-preview-31403--osdocs.netlify.app/openshift-enterprise/latest/cli_reference/openshift_cli/developer-cli-commands.html
* Admin commands: https://deploy-preview-31403--osdocs.netlify.app/openshift-enterprise/latest/cli_reference/openshift_cli/administrator-cli-commands.html

### CODE REVIEW

File: modules/oc-by-example-content.adoc [hunk 1]
```diff
@@ -0,0 +1,4723 @@
+// NOTE: The contents of this file are auto-generated
+
+[id="openshift-cli_{context}"]
+= OpenShift CLI (oc) commands
+
+
+
+ifdef::openshift-enterprise,openshift-origin[] 
+
+== oc adm build-chain
```
Comment (jboxman) [inline modules/oc-by-example-content.adoc:10]: Should these be `[discrete]`?

File: modules/oc-by-example-content.adoc [hunk 2]
```diff
@@ -0,0 +1,4723 @@
...
+Output the inputs and dependencies of your builds
+
+.Example usage
+[source,terminal,options="nowrap"]
+----
+  # Build the dependency tree for the 'latest' tag in <image-stream>
```
Comment (jboxman) [inline modules/oc-by-example-content.adoc:16]: There's extra leading space for these; I've never done this, but maybe that doesn't matter to asciidoctor.

**Review by jboxman — COMMENTED** [review]
Using using `oc help`, it does seem to group output. It's always possible to get help at the highest level, and then move down into command groups, and then individual commands.

I wonder if that's possible here?

File: modules/oc-by-example-content.adoc (hunk 1 above)
Comment (bergerhoffer) [reply modules/oc-by-example-content.adoc:10]: No, because we want them to show up in the TOC

**Comment by bergerhoffer** [comment]
> Using using `oc help`, it does seem to group output. It's always possible to get help at the highest level, and then move down into command groups, and then individual commands.
> 
> I wonder if that's possible here?

@soltysh Any thoughts on whether it would be possible to group the CLI commands into categories like is shown using `oc help`? That is how we displayed them in the docs when we did them manually: https://docs.openshift.com/container-platform/4.7/cli_reference/openshift_cli/developer-cli-commands.html

Or if that would be too manual of a process that could get missed when a new command is added.

**Comment by bobfuru** [comment]
LGTM - thank you for this, fantastic work!

**Comment by soltysh** [comment]
> @soltysh Any thoughts on whether it would be possible to group the CLI commands into categories like is shown using `oc help`? That is how we displayed them in the docs when we did them manually: https://docs.openshift.com/container-platform/4.7/cli_reference/openshift_cli/developer-cli-commands.html

@bergerhoffer there is no direct way, since that templating mechanism is part of kubectl, not the library behind commands. Having said that, everything is possible, I can imagine 2 approaches:
1. we can try parsing the help and group automatically
2. or manually copy those groups and have a check which will fail if we find new command, this will ensure we're always up-to-date (even if some manual upgrades will be required) when updating.

**Comment by lbarbeevargas** [comment]
If grouping the commands is possible, it might be a better user experience.  Otherwise, LGTM - great job on this! 😀

**Comment by bergerhoffer** [comment]
Thanks all for the feedback so far, both here and via email! I have updated previews below (this PR source isn't updated yet though) that address the weird syntax highlighting for the example code blocks, and to have two separate files again, one for developer commands and one for admin (oc adm) commands:

* Dev preview (internal): http://file.rdu.redhat.com/~ahoffer/2021/OSDOCS-1790/OSDOCS-1790/cli_reference/openshift_cli/developer-cli-commands.html
* Admin preview (internal): http://file.rdu.redhat.com/~ahoffer/2021/OSDOCS-1790/OSDOCS-1790/cli_reference/openshift_cli/administrator-cli-commands.html

We'll also investigate whether we can group the commands into categories.

**Comment by soltysh** [comment]
This looks so good :100:

**Review by kalexand-rh — COMMENTED** [review]
Since it's auto-generated, I skimmed the new modules. It LGTM.

The context of the reference assemblies is self-evident, but it might be worth adding lead-in sentences to give some space between the initial headings and meet the guideline about assemblies having an introduction. If it's possible, I suggest amending the script in a future RFE so that it adds explicit anchors for each of the headings.

File: cli_reference/openshift_cli/administrator-cli-commands.adoc [hunk 3]
```diff
@@ -1,24 +1,19 @@
...
-= OpenShift CLI administrator commands
+= OpenShift CLI administrator command reference
 include::modules/common-attributes.adoc[]
 :context: cli-administrator-commands
 
 toc::[]
 
-// Cluster management CLI commands
-include::modules/cli-administrator-cluster-management.adoc[leveloffset=+1]
+This reference provides descriptions and example commands for OpenShift CLI (`oc`) administrator commands. Information on developer commands can be found in the xref:../../cli_reference/openshift_cli/developer-cli-commands.adoc#cli-developer-commands[OpenShift CLI developer command reference].
```
Comment (kalexand-rh) [inline cli_reference/openshift_cli/administrator-cli-commands.adoc:8]: I might tighten it a bit, like: 
```suggestion
This reference provides descriptions and example commands for OpenShift CLI (`oc`) administrator commands. For developer commands, see the xref:../../cli_reference/openshift_cli/developer-cli-commands.adoc#cli-developer-commands[OpenShift CLI developer command reference].
```

**Comment by bergerhoffer** [comment]
@sferich888 Did you want to take a look at how the addition of the auto-generated CLI reference is going to look now for 4.8+? Let me know if it looks okay to you, or if you have any feedback. Thanks!

Previews:
* Dev commands: https://deploy-preview-31403--osdocs.netlify.app/openshift-enterprise/latest/cli_reference/openshift_cli/developer-cli-commands.html
* Admin commands: https://deploy-preview-31403--osdocs.netlify.app/openshift-enterprise/latest/cli_reference/openshift_cli/administrator-cli-commands.html

**Comment by bergerhoffer** [comment]
@zhouying7780 Can you please update this PR? We're replacing the manually created oc CLI reference docs with docs that are auto generated from the openshift/oc repository. The `oc-by-example-content.adoc` and `oc-adm-by-example-content.adoc` files are what was auto generated.

Previews:
* Dev commands: https://deploy-preview-31403--osdocs.netlify.app/openshift-enterprise/latest/cli_reference/openshift_cli/developer-cli-commands.html
* Admin commands: https://deploy-preview-31403--osdocs.netlify.app/openshift-enterprise/latest/cli_reference/openshift_cli/administrator-cli-commands.html

I'll run the doc generation script against the repo again after code freeze, to see if there are any additional changes. But I'd like to get this merged in before then.

**Pull request merged**